zebra/label_manager.c

```diff
@@ -0,0 +1,295 @@
+			 strerror(errno));
+		return -1;
+	}
+	zlog_debug("%s: Label chunk response received, %d bytes", __func__,
+		   size);
+
+	/* send response back */
+	stream_copy(dst, src);
+	stream_copy(zserv->obuf, zclient->ibuf);
+	ret = writen(zserv->sock, dst->data, stream_get_endp(dst));
+	if (ret <= 0) {
+		zlog_err("%s: Error sending label chunk response back: %s",
+			 __func__, strerror(errno));
+		return -1;
+	}
+	zlog_debug("%s: Label chunk response (%d bytes) sent back", __func__,
+		   ret);
+
+	return 0;
+}
+
+/**
+ * Function to initialize zclient in case this is not an actual
+ * label manager, but just a proxy to an external one.
+ *
+ * @param master Zebra master thread
+ * @param lm_zserv_path Path to zserv socket of external label manager
+ */
+static void lm_zclient_init(struct thread_master *master, char *lm_zserv_path)
+{
+	if (lm_zserv_path)
+		zclient_serv_path_set(lm_zserv_path);
+
+	/* Set default values. */
+	zclient = zclient_new(master);
+	zclient->sock = -1;
+	if (zclient_socket_connect(zclient) < 0) {
+		zlog_err("Error connecting synchronous zclient!");
+		exit(1);
+	}
+}
+
+/**
+ * Init label manager (or proxy to an external one)
+ */
+void label_manager_init(char *lm_zserv_path, struct thread_master *master)
+{
+	/* this is an actual label manager */
+	if (!lm_zserv_path) {
+		zlog_debug("Initializing own label manager");
+		lm_is_external = false;
+		lbl_mgr.lc_list = list_new();
+		lbl_mgr.lc_list->del = delete_label_chunk;
+	} else {		/* it's acting just as a proxy */
+		zlog_debug("Initializing external label manager at %s",
+			   lm_zserv_path);
+		lm_is_external = true;
+		lm_zclient_init(master, lm_zserv_path);
+	}
+}
+
+/**
+ * Core function, assigns label cunks
+ *
+ * It first searches through the list to check if there's one available
+ * (previously released). Otherwise it creates and assigns a new one
+ *
+ * @param proto Daemon protocol of client, to identify the owner
+ * @param instance Instance, to identify the owner
+ * @param keep If set, avoid garbage collection
+ * @para size Size of the label chunk
+ * @return Pointer to the assigned label chunk
+ */
+struct label_manager_chunk *assign_label_chunk(u_char proto, u_short instance,
+					       u_char keep, uint32_t size)
+{
+	struct label_manager_chunk *lmc;
+	struct listnode *node;
+
+	node = lbl_mgr.lc_list->head;
+	/* first check if there's one available */
+	for (ALL_LIST_ELEMENTS_RO(lbl_mgr.lc_list, node, lmc)) {
+		if (lmc->proto == NO_PROTO && lmc->end - lmc->start + 1 == size) {
+			lmc->proto = proto;
+			lmc->instance = instance;
+			lmc->keep = keep;
+			return lmc;
+		}
+	}
+	/* otherwise create a new one */
+	lmc = XCALLOC(MTYPE_LM_CHUNK, sizeof(struct label_manager_chunk));
+
+	if (list_isempty(lbl_mgr.lc_list))
+		lmc->start = MPLS_MIN_UNRESERVED_LABEL;
+	else
+		lmc->start = ((struct label_manager_chunk *)
+			      listgetdata(listtail(lbl_mgr.lc_list)))->end + 1;
+	if (lmc->start > MPLS_MAX_UNRESERVED_LABEL - size + 1) {
```

Comment:
Well, adding the possibility to specify the beginning of the label range in the request for Segment Routing completely break the way you manage chunk. In fact, there is a risk that we got hole in the list of allocated range.